### PR TITLE
fix before_remove and after_remove relation callbacks and add shift r…

### PR DESCRIPTION
**Pull request by dks17**
When object of a model has `embeds_many` relation with `before_remove` or `after_remove` relation callbacks. Callback does not work when `shift` method is called on `embeds_many` relation. This pr fix it.

### CODE REVIEW

File: lib/mongoid/association/embedded/embeds_many/proxy.rb [hunk 1]
```diff
@@ -271,6 +271,29 @@ def pop(count = nil)
...
+          #
+          # @return [ Document, Array<Document> ] The shifted document(s).
+          def shift(count = nil)
+            if count
+              if _target.size > 0 && docs = _target[0, count]
+                docs.each { |doc| delete(doc) }
+              end
+            else
+              delete(_target[0])
+            end
```
Comment (estolfo) [inline lib/mongoid/association/embedded/embeds_many/proxy.rb]: We can make this the following, for simplicity:

```ruby
  shifted = _target[0, count || 1].each { |doc| delete(doc) }
  shifted.size <= 1 ? shifted[0] : shifted
```

File: spec/mongoid/association/embedded/embeds_many/proxy_spec.rb [hunk 2]
```diff
@@ -2601,6 +2601,108 @@ class TrackingIdValidationHistory
...
+
+      let!(:address_two) do
+        person.addresses.create(street: "hermannstr")
+      end
+
+      context "when the number is not larger than the relation" do
+
+        let!(:shifted) do
+          # person.addresses.create(street: "kos")
+          # binding.pry
```
Comment (estolfo) [inline spec/mongoid/association/embedded/embeds_many/proxy_spec.rb]: Can we remove these commented lines please?

File: lib/mongoid/association/embedded/embeds_many/proxy.rb (hunk 1 above)
Comment (dks17) [reply lib/mongoid/association/embedded/embeds_many/proxy.rb]: I have added test when the number is zero and this implementation fails the tests.

Comment (estolfo) [reply lib/mongoid/association/embedded/embeds_many/proxy.rb]: Ok, thank you for adding that edge case test. Your implementation is good then, I'll merge this, thank you!
